src/inmanta/protocol/openapi/converter.py

```diff
@@ -0,0 +1,185 @@
+"""
+    Copyright 2020 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+import json
+from typing import Dict, List, Type
+
+from pydantic.main import BaseModel
+
+from inmanta.protocol.common import ArgOption, UrlMethod
+from inmanta.protocol.openapi.model import OpenAPI, Operation, Parameter, ParameterType, PathItem, RequestBody, Response
+from inmanta.server import config
+from inmanta.util import get_compiler_version
+
+
+def openapi_json_encoder(o):
+    if isinstance(o, BaseModel):
+        return o.dict(by_alias=True, exclude_none=True)
+
+
+class OpenApiConverter:
+    """
+        Extracts API information for the OpenAPI definition from the server
+    """
+
+    def get_servers(self):
+        bind_port = config.get_bind_port()
+        bind_addresses = config.server_bind_address.get()
+        return [{"url": f"http://{bind_address}:{bind_port}/"} for bind_address in bind_addresses]
+
+    def get_top_level_object(self):
+        info = {"title": "Inmanta Service Orchestrator", "version": get_compiler_version()}
+        servers = self.get_servers()
+        return {"openapi": "3.0.2", "info": info, "paths": {}, "servers": servers}
+
+    def filter_api_methods(self, methods: Dict[str, UrlMethod]) -> Dict[str, UrlMethod]:
+        return {
+            method_name: url_method
+            for method_name, url_method in methods.items()
+            if "api" in url_method.properties.client_types
+        }
+
+    def get_path_of_method(self, methods: Dict[str, UrlMethod]) -> str:
+        path = [
+            f"/{method.properties._api_prefix}/v{method.properties.api_version}{method.properties._path.path}".replace(
+                "<", "{"
+            ).replace(">", "}")
+            for method in methods.values()
+        ][0]
+        return path
+
+    def extract_parameters_from_arg_options(self, response: Response, path: str, arg_options: Dict[str, ArgOption]):
+        parameters = []
+        for option_name, option in arg_options.items():
+            param_dict = {}
+            if option.header:
+                param_dict["in"] = ParameterType.header
+                param_dict["name"] = option.header
+                param_dict["schema"] = {"type": "string"}
+                if option.reply_header:
+                    response.headers = {option.header: {"description": option.header, "schema": {"type": "string"}}}
```

Comment:
no name field here? Why is the format so different compared to the param_dict?

Reply:
The name is the key in this dict:
https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#header-object
So a Header is almost the same as a Parameter, but not exactly